### PR TITLE
Job execution modal

### PR DESCRIPTION
**Pull request by smk4664**
# NAPPS-925

# What's Changed

This PR introduces a new Detail Component from which we can launch Jobs in Nautobot. The component has the ability to customize the fields that are displayed in the job_form, prefill data into the job_form and run custom Javascript in the job_form.  This utilizes HTMX to render the modal and swap the content in the modal. There is a modification in the JobRunView to support the new modal, and a new view action on the Job Result to return a job result modal. 

# Screenshots

image

image

image

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/core/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/core/templates/components/button/open_modal.html [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<button 
+    type="button"
+    class="btn btn-{{ color }}{% if size %} btn-{{ size }}{% endif %}"
+    {% for key, value in attributes.items %}{{ key }}="{{ value }}" {% endfor %}
+    data-bs-toggle="modal"
+    data-bs-target="#{{ modal_id }}"
+    hx-get="{{ hx_get }}"
+    hx-target="#modal-content-container"
```
Comment (glennmatthews) [inline nautobot/core/templates/components/button/open_modal.html]: Possibly might be desirable for this to be another parameter to the component and render context rather than being hard-coded?

File: nautobot/core/ui/object_detail.py [hunk 2]
```diff
@@ -2608,3 +2609,97 @@ def render_label(self, context: Context):
...
+        return {
+            **super().get_extra_context(context),
+            "modal_id": self.modal_id,
+            "hx_get": self.hx_get,
+            "hx_select": self.hx_select,
+            "hx_select_oob": self.hx_select_oob,
+            "hx_swap": self.hx_swap,
+            "hx_target": self.hx_target,
+            "modal_title": self.modal_title,
+            "hx_vals": json.dumps(self.hx_vals),
```
Comment (glennmatthews) [inline nautobot/core/ui/object_detail.py]: Wonder if it might be cleaner and safer to use the `json_script` template tag instead of `json.dumps` + `safe` (https://docs.djangoproject.com/en/5.2/ref/templates/builtins/#json-script)

File: nautobot/core/templates/components/button/open_modal.html [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    hx-vals='{{ hx_vals|safe }}'
+    {% if hx_swap %}hx-swap="{{ hx_swap }}"{% endif %}
+    {% if hx_select %}hx-select="{{ hx_select }}"{% endif %}
+    {% if hx_select_oob %}hx-select-oob="{{ hx_select_oob }}"{% endif %}
+    hx-on:click="document.getElementById('modal-fallback-title').innerText = '{{ modal_title }}'"
```
Comment (glennmatthews) [inline nautobot/core/templates/components/button/open_modal.html]: Probably good to parameterize `modal-fallback-title` as well.

File: nautobot/core/templates/inc/generic_modal.html [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+                <h4 class="modal-title" id="modal-fallback-title">
+                    Loading...
+                </h4>
+                <button
+                    type="button"
+                    class="btn-close"
+                    data-bs-dismiss="modal"
+                    aria-label="Close"
+                ></button>
+            </div>
```
Comment (glennmatthews) [inline nautobot/core/templates/inc/generic_modal.html]: Can/should we use `inc/modal_header.html` here?

File: nautobot/core/ui/object_detail.py [hunk 5]
```diff
@@ -2608,3 +2609,97 @@ def render_label(self, context: Context):
...
+        **kwargs,
+    ):
+        self.modal_id = modal_id
+        self.hx_get = hx_get
+        self.hx_select = hx_select
+        self.hx_select_oob = hx_select_oob
+        self.hx_swap = hx_swap
+        self.hx_target = hx_target
+        self.modal_title = modal_title
+        self.hx_vals = hx_vals or {}
```
Comment (glennmatthews) [inline nautobot/core/ui/object_detail.py]: Let's follow the newer component pattern that Gary introduced in #8533.

File: nautobot/core/ui/object_detail.py [hunk 6]
```diff
@@ -2608,3 +2609,97 @@ def render_label(self, context: Context):
...
+            "hx_select": self.hx_select,
+            "hx_select_oob": self.hx_select_oob,
+            "hx_swap": self.hx_swap,
+            "hx_target": self.hx_target,
+            "modal_title": self.modal_title,
+            "hx_vals": json.dumps(self.hx_vals),
+        }
+
+
+def resolve_attr(obj, dotted_path: str):
```
Comment (glennmatthews) [inline nautobot/core/ui/object_detail.py:2617]: Maybe this belongs in nautobot.core.utils.lookup?

File: nautobot/core/ui/object_detail.py [hunk 7]
```diff
@@ -2608,3 +2609,97 @@ def render_label(self, context: Context):
                 badge(get_obj_from_context(context).associated_object_metadata.count(), True),
             ),
         )
+
+
+class ButtonOpenModal(Button):
```
Comment (glennmatthews) [inline nautobot/core/ui/object_detail.py]: Perhaps `ModalOpeningButton` might be clearer?

File: nautobot/core/ui/object_detail.py [hunk 8]
```diff
@@ -2608,3 +2609,97 @@ def render_label(self, context: Context):
...
+        modal_id="nautobot-generic-modal",
+        hx_get="",
+        hx_select=None,
+        hx_select_oob=None,
+        hx_target=None,
+        hx_vals=None,
+        hx_swap="innerHTML",
+        modal_title="",
+        **kwargs,
+    ):
```
Comment (glennmatthews) [inline nautobot/core/ui/object_detail.py]: Please provide docstring with (ideally) examples :-)

File: nautobot/extras/templates/extras/htmx/job_form_modal.html [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+{% include 'inc/modal_header.html' with title=title %}
+
+<form hx-post="{% url 'extras:job_run_by_class_path' class_path=class_path %}" 
+      hx-target="#modal-content-container" 
+      id="job-form-modal"
+      hx-vals='{{ hx_vals|safe }}'
+      enctype="multipart/form-data">
+    {% csrf_token %}
+    <div class="modal-body" style="max-height: 70vh; overflow-y: auto;">
+        {% render_form job_form excluded_fields=excluded_json|safe %}
```
Comment (glennmatthews) [inline nautobot/extras/templates/extras/htmx/job_form_modal.html]: Feels off to me that we're passing the excluded fields as a JSON string instead of the underlying Python object?

File: nautobot/extras/templates/extras/htmx/job_form_modal.html [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        <div class="form-check mb-10 d-md-flex justify-content-center">
+            <div class="col-md-9 offset-md-3">
+                <div class="form-check mt-6">
+                    <input class="form-check-input" type="checkbox" id="toggle-advanced">
+                    <label class="form-check-label" for="toggle-advanced">Show Advanced Settings</label>
+                </div>
+            </div>
+        </div>
+        <div id="advanced-settings" class="d-none">
+            {% if advanced_fields %}
```
Comment (glennmatthews) [inline nautobot/extras/templates/extras/htmx/job_form_modal.html]: Is the `if` condition needed around the `for` loop? Or can we just have `advanced_fields` be `[]` and make the for loop be a no-op?

File: nautobot/extras/templates/extras/jobresult_modal.html [hunk 11]
```diff
@@ -0,0 +1,58 @@
+<div class="modal-header">
+    <h4 class="modal-title">Job Result: {{ result.job_model.name|default:result.name }}</h4>
+    <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
+</div>
```
Comment (glennmatthews) [inline nautobot/extras/templates/extras/jobresult_modal.html]: Can we use inc/modal_header.html?

File: nautobot/extras/templates/extras/jobresult_modal.html [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+<div class="modal-body">
+    {# Polling Logic: Only present if the job is not finished #}
+    <div id="job-polling-container"
+        {% if result.status != 'FAILURE' and result.status != 'REVOKED' and result.status != 'SUCCESS' %}
```
Comment (glennmatthews) [inline nautobot/extras/templates/extras/jobresult_modal.html]: If we don't have an API or property already on `JobResult` to implement this logic, we should add one.

File: nautobot/extras/templates/extras/jobresult_modal.html [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+            <tbody>
+                <tr>
+                    <td><strong>Job Status</strong></td>
+                    <td>
+                        <span id="pending-result-label">{% include 'extras/inc/job_label.html' with result=result %}</span>
+                    </td>
+                </tr>
+                <tr>
+                    <td><strong>Duration</strong></td>
+                    <td>{{ result.duration|default:"--" }}</td>
```
Comment (glennmatthews) [inline nautobot/extras/templates/extras/jobresult_modal.html]: Either use `placeholder` templatetag here instead of `default`, or have it show a bootstrap spinner instead like the JobResult detail view does.

File: nautobot/core/ui/object_detail.py (hunk 2 above)
Comment (smk4664) [reply nautobot/core/ui/object_detail.py]: I tried the json_script template tag, but that automatically inserts the <script> tab.

File: nautobot/core/ui/object_detail.py (hunk 6 above)
Comment (smk4664) [reply nautobot/core/ui/object_detail.py:2617]: Agreed, stuck here for initial development.

File: nautobot/extras/templates/extras/jobresult_modal.html (hunk 12 above)
Comment (smk4664) [reply nautobot/extras/templates/extras/jobresult_modal.html]: This was previously handled by the javascript.

File: nautobot/core/tests/test_ui.py [hunk 14]
```diff
@@ -799,3 +801,64 @@ def test_tab_conditional_rendering(self):
...
+        self.assertEqual(hx_vals["run_button_label"], "Execute!")
+        self.assertEqual(hx_vals["job_result_key"], "output_data")
+
+        # Verify Bootstrap and HTMX target attributes
+        self.assertEqual(btn.attributes["data-bs-toggle"], "modal")
+        self.assertEqual(btn.attributes["data-bs-target"], "#nautobot-generic-modal")
+        self.assertEqual(btn.attributes["hx-target"], "#modal-content-container")
+
+        # Verify URL generation
+        expected_url = reverse("extras:job_run_by_class_path", kwargs={"class_path": "nautobot.core.jobs.SampleJob"})
```
Comment (glennmatthews) [inline nautobot/core/tests/test_ui.py:876]: Feels like somewhere in `get_extra_context` we should probably validate that the `class_path` refers to an actual existing Job database record and maybe render something different (like maybe a disabled button) if not?

File: nautobot/core/ui/object_detail.py [hunk 15]
```diff
@@ -2608,3 +2609,94 @@ def render_label(self, context: Context):
...
+
+    class_path = None
+    advanced_fields = []
+    initial_field_mapping = {}
+    run_button_label = "Run Job Now"
+    job_result_key = None
+
+    def __init__(self, **kwargs):
+        """
+        Initialize a Button component.
```
Comment (glennmatthews) [inline nautobot/core/ui/object_detail.py]:
```suggestion
        Initialize a _JobModalButton component.
```

File: nautobot/core/ui/object_detail.py [hunk 16]
```diff
@@ -2608,3 +2609,94 @@ def render_label(self, context: Context):
...
+    except (AttributeError, ObjectDoesNotExist):
+        return None
+
+
+class _JobModalButton(Button):
+    """A Button that, when clicked, opens a modal dialog for running a Job."""
+
+    class_path = None
+    advanced_fields = []
+    initial_field_mapping = {}
```
Comment (glennmatthews) [inline nautobot/core/ui/object_detail.py]: I have a slight concern about these being mutable objects? `advanced_fields` should probably default to `()` at least... actually it looks like there's a number of other examples of this antipattern in the module already, though I won't ask you to do anything about them in this PR. :-)

File: nautobot/extras/templates/extras/htmx/job_form_modal.html [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+                    const dryrunInput = container.querySelector('#id_dryrun');
+                    if (!dryrunInput) return;
+
+                    const isDryRun = dryrunInput.type === 'checkbox' ? dryrunInput.checked : dryrunInput.value === 'True';
+                    const runButton = container.querySelector('#id__run');
+                    
+                    if (runButton) {
+                        const text = `Run Job Now${isDryRun ? ' (DRYRUN)' : ''}`;
+                        // Replacing the last child node (the text) safely
+                        runButton.childNodes[runButton.childNodes.length - 1].replaceWith(text);
```
Comment (glennmatthews) [inline nautobot/extras/templates/extras/htmx/job_form_modal.html]: Doesn't this conflict with the customizable `run_button_label`?

File: nautobot/extras/templates/extras/jobresult_modal.html [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+            <div class="row align-items-center" style="margin-bottom: 12px;">
+                <div class="col-4 text-end fw-semibold">Job Status:</div>
+                <div class="col-8">
+                    {% include 'extras/inc/job_label.html' with result=result %}
+                </div>
+            </div>
+            <div class="row align-items-center" style="margin-bottom: 12px;">
+                <div class="col-4 text-end fw-semibold">Duration:</div>
+                <div class="col-8">
+                    <span>{{ result.duration|default:"--" }}</span>
```
Comment (glennmatthews) [inline nautobot/extras/templates/extras/jobresult_modal.html]: `placeholder` or Bootstrap spinner here for the still-running case, please.

File: nautobot/extras/views.py [hunk 19]
```diff
@@ -2394,16 +2467,51 @@ def post(self, request, class_path=None, pk=None):
...
-            request,
-            template_name,
-            {
-                "job_model": job_model,
-                "job_form": job_form,
-                "job_execution_form": job_execution_form,
-                "schedule_form": schedule_form,
-            },
-        )
+        if htmx_request and htmx_modal:
```
Comment (glennmatthews) [inline nautobot/extras/views.py]: If this is the same as lines 2284-2327 above, can we create a common helper function to reduce duplication of code?

File: nautobot/extras/views.py [hunk 20]
```diff
@@ -3065,6 +3173,32 @@ def _handle_console_poll(self, request, job_result) -> HttpResponse:
...
+    def modal(self, request, *args, **kwargs):
+        instance = self.get_object()
+        title = instance.job_model.name
+        job_result_key = request.GET.get("job_result_key", None)
+        detail_value = f"Job finished with status: {instance.get_status_display()}"
+        if instance.result and isinstance(instance.result, dict) and job_result_key:
+            detail_value = instance.result.get(
+                job_result_key, f"Job finished with status: {instance.get_status_display()}"
+            )
+        elif instance.result and isinstance(instance.result, (str, int, float)):
```
Comment (glennmatthews) [inline nautobot/extras/views.py]: Should this just be `elif instance.result is not None`?

Comment (smk4664) [reply nautobot/extras/views.py]: I don't believe so, if it is a dictionary, but no key is used, it would print the dictionary in the Details box.  But, I could change that.

Comment (glennmatthews) [reply nautobot/extras/views.py]: Ah good point. Not sure what the preferred behavior would be in that case.

Comment (glennmatthews) [reply nautobot/extras/views.py]: But maybe at least consider the possibility of `result == 0` as something that should be displayed?

File: nautobot/extras/templates/extras/htmx/job_form_modal.html (hunk 17 above)
Comment (smk4664) [reply nautobot/extras/templates/extras/htmx/job_form_modal.html]: Yes... I will fix, thanks for checking this. I am going to set an id for that button.

File: nautobot/core/templates/generic/object_retrieve.html [hunk 21]
```diff
@@ -318,4 +319,50 @@
 
         window.addEventListener('popstate', (event) => show_tab(event.state.id));
     </script>
+    <script>
+        document.body.addEventListener('htmx:afterSwap', (event) => {
+            // Check if the swap happened inside our shared modal
+            if (event.detail.target.id === "modal-content-container") {
```
Comment (bernalvinicius) [inline nautobot/core/templates/generic/object_retrieve.html]: Potential runtime issue: `event.detail.target` may be `undefined` in some edge cases, which would throw here. Consider using `event.detail?.target?.id === "modal-content-container"` to avoid breaking the script.

File: nautobot/core/tests/test_ui.py (hunk 14 above)
Comment (smk4664) [reply nautobot/core/tests/test_ui.py:876]: Good point, let me try to add this in.

File: nautobot/core/ui/object_detail.py (hunk 16 above)
Comment (smk4664) [reply nautobot/core/ui/object_detail.py]: Resolved

File: nautobot/extras/views.py (hunk 20 above)
Comment (smk4664) [reply nautobot/extras/views.py]: Great feedback! I am going to go ahead and try to display all values, but fall back to the "Job finished with status:"

File: nautobot/core/tests/test_ui.py (hunk 14 above)
Comment (glennmatthews) [reply nautobot/core/tests/test_ui.py:876]: Looks good - maybe add a corresponding unit test as well?

File: nautobot/core/ui/object_detail.py [hunk 22]
```diff
@@ -2610,3 +2612,113 @@ def render_label(self, context: Context):
...
+            }
+        )
+        # If the user doesn't have permission to the Job, or the Job doesn't exist, disable the button.
+        try:
+            jobs = Job.objects
+            if "request" in context and context["request"].user is not None:
+                jobs = jobs.restrict(context["request"].user, "view")
+            jobs.get_for_class_path(self.class_path)
+        except Job.DoesNotExist:
+            self.attributes["disabled"] = "disabled"
```
Comment (glennmatthews) [inline nautobot/core/ui/object_detail.py:2723]: Actually, thinking about this again, would it be cleaner to do this check in `should_render()` and just not even render the button in such a case? Would align with how we don't render `edit` or `delete` buttons if the user lacks those permissions. Up to you.

File: nautobot/extras/views.py [hunk 23]
```diff
@@ -2200,7 +2210,95 @@ def _handle_immediate_execution(
...
+                    self.request, f'Unable to render requested custom job template "{job_class.template_name}": {err}'
+                )
+        return template_name
+
+    def _render_response(self, request, job_model, job_class, job_form, job_execution_form, schedule_form):
+        """Helper function to render the appropriate response, including handling HTMX modals."""
+        htmx_request = self.request.headers.get("HX-Request", False)
+        htmx_modal = False
+        title = "Run Job"
+        if hasattr(job_model, "name"):
```
Comment (glennmatthews) [inline nautobot/extras/views.py]: Would this be clearer and still correct if written as `if job_model is not None`?

File: nautobot/extras/views.py [hunk 24]
```diff
@@ -2200,7 +2210,95 @@ def _handle_immediate_execution(
...
+            title = job_model.name
+        run_button_label = "Run Job Now"
+        job_result_key = None
+        advanced_fields = ()
+        if htmx_request:
+            if request.method == "POST":
+                htmx_modal = request.POST.get("job_form_modal", False)
+                run_button_label = request.POST.get("run_button_label", "Run Job Now")
```
Comment (glennmatthews) [inline nautobot/extras/views.py:2248]: Feels a little bit like `run_button_label` could be implemented as another optional param on the `job_class` instead of passing it around as a request parameter. (Guess that wouldn't allow for different JobModalButton labels for a single Job, but not sure that's really a relevant use case anyways...) Not blocking.

File: nautobot/extras/views.py [hunk 25]
```diff
@@ -2200,7 +2210,95 @@ def _handle_immediate_execution(
...
+                job_result_key = request.POST.get("job_result_key", None)
```
Comment (glennmatthews) [inline nautobot/extras/views.py:2249]: Same thought on `job_result_key` and even `advanced_field_names` as well, actually.

File: nautobot/extras/views.py [hunk 26]
```diff
@@ -2237,34 +2335,20 @@ def get(self, request, class_path=None, pk=None):
...
-                    get_template(job_class.template_name)
-                    template_name = job_class.template_name
-                except TemplateDoesNotExist as err:
-                    messages.error(
-                        request, f'Unable to render requested custom job template "{job_class.template_name}": {err}'
-                    )
         except RuntimeError as err:
             messages.error(request, f"Unable to run or schedule '{job_model}': {err}")
+            if htmx_request and htmx_modal:
+                return render(request, "extras/htmx/job_missing_modal.html", {"class_path": class_path})
```
Comment (glennmatthews) [inline nautobot/extras/views.py]: Should we skip the `messages.error()` call in the htmx case?

File: nautobot/extras/views.py [hunk 27]
```diff
@@ -2237,34 +2335,20 @@ def get(self, request, class_path=None, pk=None):
...
-            template_name,  # 2.0 TODO: extras/job_submission.html
-            {
-                "job_model": job_model,
-                "job_form": job_form,
-                "job_execution_form": job_execution_form,
-                "schedule_form": schedule_form,
-            },
-        )
+        response = self._render_response(request, job_model, job_class, job_form, job_execution_form, schedule_form)
+        patch_vary_headers(response, ["HX-Request"])
```
Comment (glennmatthews) [inline nautobot/extras/views.py]: Might as well move this line inside `_render_response`?

File: nautobot/extras/views.py (hunk 23 above)
Comment (smk4664) [reply nautobot/extras/views.py]: I think there should always be a job_model, but seems we fake it during tests. I think what really is happening is the name can be blank.

File: nautobot/extras/views.py (hunk 24 above)
Comment (smk4664) [reply nautobot/extras/views.py:2248]: Good question... I tried to put as much as I could on the JobButton and left the Job alone, but that does mean passing quite a bit in the request.

File: nautobot/extras/views.py (hunk 25 above)
Comment (smk4664) [reply nautobot/extras/views.py:2249]: The advanced_field_names were specifically called out as being part of the Component during the design call.

File: nautobot/extras/views.py (hunk 27 above)
Comment (smk4664) [reply nautobot/extras/views.py]: I thought about it, and since you mention it, I will.
